File: url/creator.go
package url

import (
	"fmt"
	"math/rand"
	"net/url"
	"strconv"
	"time"
)

// Creator creates URLs using pre defined heuristics and the provided params
type Creator interface {
	Create(params interface{}) (string, error)
}

func init() {
	rand.Seed(time.Now().UnixNano())
}

const (
	request = "https://www.googleapis.com/customsearch/v1"

	imageSearchType = "image"
)

// GoogleImagesCreator creates URLs for querying Google images
type GoogleImagesCreator struct {
}

// GoogleImagesParams is the struct providing the params for the GoogleImagesCreator
type GoogleImagesParams struct {
	APIKey     string
	APICX      string
	BaseQuery  string
	ImgSize    string
	Randomness int
	ResultNum  int
	Features   []string
}

// Create creates a Google Images url using a set of URL parameters
func (g GoogleImagesCreator) Create(rawParams interface{}) (string, error) {
	gparams, ok := rawParams.(GoogleImagesParams)
	if !ok {
		return "", fmt.Errorf("GoogleImagesCreator must take a GoogleImagesParams params")
	}

	baseURL, err := url.Parse(request)
	if err != nil {
		panic("invalid base request url")
	}

	params := &url.Values{}
	baseURL.RawQuery = g.addParams(params, gparams)

	return baseURL.String(), nil

}

func (g GoogleImagesCreator) addParams(params *url.Values, gparams GoogleImagesParams) string {
	params.Add("key", gparams.APIKey)
	params.Add("cx", gparams.APICX)
	params.Add("q", gparams.BaseQuery+" "+gparams.Features[rand.Intn(len(gparams.Features))])
	params.Add("imgSize", gparams.ImgSize)
	params.Add("start", strconv.Itoa(rand.Intn(gparams.Randomness)))
	params.Add("searchType", imageSearchType)
	params.Add("num", strconv.Itoa(gparams.ResultNum))

	return params.Encode()
}

File: typescript/src/url/creator.ts
export interface Creator {
  create(params: any): Promise<string>;
}

export interface GoogleImagesParams {
  apiKey: string;
  apiCX: string;
  baseQuery: string;
  imgSize: string;
  randomness: number;
  resultNum: number;
  features: string[];
}

export class GoogleImagesCreator implements Creator {
  private static readonly REQUEST_URL = 'https://www.googleapis.com/customsearch/v1';
  private static readonly IMAGE_SEARCH_TYPE = 'image';

  async create(rawParams: any): Promise<string> {
    const params = rawParams as GoogleImagesParams;
    
    if (!params.apiKey || !params.apiCX) {
      throw new Error('GoogleImagesCreator requires apiKey and apiCX');
    }

    const baseURL = new URL(GoogleImagesCreator.REQUEST_URL);
    const searchParams = this.addParams(params);
    
    searchParams.forEach((value, key) => {
      baseURL.searchParams.append(key, value);
    });

    return baseURL.toString();
  }

  private addParams(params: GoogleImagesParams): URLSearchParams {
    const searchParams = new URLSearchParams();
    
    searchParams.set('key', params.apiKey);
    searchParams.set('cx', params.apiCX);
    
    const randomFeature = params.features[Math.floor(Math.random() * params.features.length)];
    searchParams.set('q', `${params.baseQuery} ${randomFeature}`);
    
    if (params.imgSize !== 'all') {
      searchParams.set('imgSize', params.imgSize);
    }
    
    searchParams.set('start', Math.floor(Math.random() * params.randomness).toString());
    searchParams.set('searchType', GoogleImagesCreator.IMAGE_SEARCH_TYPE);
    searchParams.set('num', params.resultNum.toString());

    return searchParams;
  }
}